pkg/adapter/nimbus-k8tls/manager/manager.go

```diff
@@ -0,0 +1,350 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 Authors of Nimbus
+
+package manager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	intentv1 "github.com/5GSEC/nimbus/api/v1"
+	"github.com/5GSEC/nimbus/pkg/adapter/common"
+	"github.com/5GSEC/nimbus/pkg/adapter/k8s"
+	"github.com/5GSEC/nimbus/pkg/adapter/nimbus-k8tls/builder"
+	"github.com/5GSEC/nimbus/pkg/adapter/nimbus-k8tls/watcher"
+	adapterutil "github.com/5GSEC/nimbus/pkg/adapter/util"
+	globalwatcher "github.com/5GSEC/nimbus/pkg/adapter/watcher"
+)
+
+var (
+	scheme        = runtime.NewScheme()
+	k8sClient     client.Client
+	NamespaceName = "nimbus-k8tls-env"
+)
+
+func init() {
+	utilruntime.Must(corev1.AddToScheme(scheme))
+	utilruntime.Must(batchv1.AddToScheme(scheme))
+	utilruntime.Must(rbacv1.AddToScheme(scheme))
+	utilruntime.Must(intentv1.AddToScheme(scheme))
+	k8sClient = k8s.NewOrDie(scheme)
+}
+
+//+kubebuilder:rbac:groups=intent.security.nimbus.com,resources=clusternimbuspolicies,verbs=get;list;watch
+//+kubebuilder:rbac:groups=intent.security.nimbus.com,resources=clusternimbuspolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;create;delete;list;watch;update
+//+kubebuilder:rbac:groups=batch,resources=cronjobs,verbs=get;create;delete;list;watch;update
+
+func Run(ctx context.Context) {
+	cwnpCh := make(chan string)
+	deletedcwnpCh := make(chan string)
+	go globalwatcher.WatchClusterNimbusPolicies(ctx, cwnpCh, deletedcwnpCh)
+
+	updateJobCh := make(chan common.Request)
+	deletedJobCh := make(chan common.Request)
+	go watcher.WatchJobs(ctx, updateJobCh, deletedJobCh)
+
+	updateCronJobCh := make(chan common.Request)
+	deletedCronJobCh := make(chan common.Request)
+	go watcher.WatchCronJobs(ctx, updateCronJobCh, deletedCronJobCh)
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(cwnpCh)
```

Comment:
Since we're already passing parent context to informers, they will also be stopped when the cancellation signal is triggered.

https://github.com/5GSEC/nimbus/blob/4aec2fc85fbfda9ca716f86cf4a9338b56ab0901/pkg/adapter/watcher/nimbuspolicy_watcher.go#L108

https://github.com/kubernetes/client-go/blob/8e3349bf54992d5a0222fe8512b07595daf0ecdf/informers/factory.go#L250

Reply:
That is correct, but we cannot ensure that the informer will be stopped before the manager closes the channel.

Reply:
https://github.com/5GSEC/nimbus/issues/187